firedrake/cython/dmswarm.pyx

```diff
@@ -0,0 +1,508 @@
+        ghost_cell_indices = np.full(ncells, -1, dtype=IntType)
+
+        # Search for ghost cell indices (spooky!)
+        sf = plex.getPointSF()
+        CHKERR(PetscSFGetGraph(sf.sf, &nroots, &nleaves, &ilocal, &iremote))
+        for i in range(nleaves):
+            if cStart <= ilocal[i] < cEnd:
+                # NOTE need to check this is correct index. Can I check the labels some how?
+                ghost_cell_indices[ilocal[i] - cStart] = ilocal[i]
+
+        # trim -1's to reduce searching needed
+        ghost_cell_indices = ghost_cell_indices[ghost_cell_indices != -1]
+
+        # remove swarm pic parent cell indices which match ghost cell indices
+        for i in range(npics-1, -1, -1):
+            if np.isin(pic_cell_indices[i], ghost_cell_indices):
+                # removePointAtIndex shift cell numbers down by 1
+                swarm.removePointAtIndex(i)
+
+    return
+
+@cython.boundscheck(False)
+@cython.wraparound(False)
+def label_pic_parent_cell_nums(PETSc.DM swarm, parentmesh):
+    """
+    For each PIC in the input swarm, label its `parentcellnum` field with
+    the relevant cell number from the `parentmesh` in which is it emersed.
+    The cell numbering is that given by the `parentmesh.locate_cell`
+    method.
+
+    :arg swarm: The DMSWARM which contains the PICs immersed in
+        `parentmesh`
+    :arg parentmesh: The mesh within with the `swarm` PICs are immersed.
+
+    ..note:: All PICs must be within the parentmesh or this will try to
+             assign `None` (returned by `parentmesh.locate_cell`) to a
+             `PetscReal`.
+    """
+    cdef:
+        PetscInt num_vertices, i, dim
+        PetscReal parent_cell_num
+        np.ndarray[PetscReal, ndim=2, mode="c"] swarm_coords
+        np.ndarray[PetscReal, ndim=1, mode="c"] parent_cell_nums
+
+    dim = parentmesh.geometric_dimension()
+
+    num_vertices = swarm.getLocalSize()
+
+    # Check size of biggest num_vertices so
+    # locate_cell can be called on every processor
+    comm = swarm.comm.tompi4py()
+    max_num_vertices = comm.allreduce(num_vertices, op=MPI.SUM)
+
+    # Create an out of mesh point to use in locate_cell when needed
+    out_of_mesh_point = np.empty(shape=(1, dim))
+    out_of_mesh_point.fill(np.inf)
+
+    # get fields - NOTE aren't copied so could have GC issues!
+    swarm_coords = swarm.getField("DMSwarmPIC_coor").reshape((num_vertices, dim))
+    parent_cell_nums = swarm.getField("parentcellnum")
+
+    # find parent cell numbers
+    for i in range(max_num_vertices):
+        if i < num_vertices:
+            parent_cell_num = parentmesh.locate_cell(swarm_coords[i])
+            parent_cell_nums[i] = parent_cell_num
+        else:
+            parentmesh.locate_cell(out_of_mesh_point)  # should return None
+
+    # have to restore fields once accessed to allow access again
+    swarm.restoreField("parentcellnum")
+    swarm.restoreField("DMSwarmPIC_coor")
+
+    return
```

Comment:
No need to return at end of functions with no return value. Also please add a newline at end of file

Reply:
Changed